docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,407 @@
+
+- *User*, who uses NKG. For logging, the user persona maps to the cluster operator from
+  the [Gateway API personas](https://gateway-api.sigs.k8s.io/concepts/security-model/#roles-and-personas).
+- *Developer*, who develops NKG.
+
+Below are common user stories for both personas and also persona-specific stories.
+
+#### Common Stories
+
+As a user/developer, I want to:
+
+- Have confidence the control plane works properly by looking at the logs.
+- Control the amount/level of detail of logged information.
+- Use logs to troubleshoot an issue with the control plane, either in real-time or by looking at the log history.
+- In case of errors, see what event or Kubernetes resource caused it.
+
+#### Stories For User
+
+As a user, I want to:
+
+- Export the logs into my logging infrastructure and parse them for further processing (like querying).
+- Share the logs with developers to ask for help or report a bug.
+
+#### Stories For Developers
+
+As a developer, I want to:
+
+- See human-readable (text) as opposed to machine-readable (JSON) format.
+- Use logs to troubleshoot an issue collected from a user.
+
+## Logging Library Choice
+
+To make the logs easily parseable/processable by other programs, we use structured logging - i.e. logging data in a
+structured
+way (like JSON) as opposed to a non-structured way (like a long string from a `printf` invocation).
+
+The code logs via the [logr][logr] library, which abstracts away
+the logger implementation. This library covers our needs, is well adopted, and many logger implementations implement its
+interfaces.
+
+As an implementation, we use [zap](https://github.com/uber-go/zap/), which:
+
+- supports logr via https://github.com/go-logr/zapr
+- supports both JSON and text format
+- supports log levels
+- demonstrates good [performance](https://github.com/uber-go/zap/#performance)
+- is stable and well-adopted.
+
+We initialize zap via controller-runtime
+framework [helpers][cr-zap].
+
+## Guidelines
+
+### How to Log
+
+#### Log Levels
+
+The table below describes the log verbosity levels. A level includes all messages of the levels above.
+
+| Level            | Contents                                                                             | Target persona |
+|------------------|--------------------------------------------------------------------------------------|----------------|
+| `error`          | errors                                                                               | user           |
+| `info` (default) | information about how the control plane operates                                     | user           |
+| `debug`          | extra information about how the control plane operates to help troubleshoot an issue | developer      |
+
+> The levels match the levels of
+> the [controller-runtime framework][cr-log-levels] without any custom levels.
+
+#### Log Messages
+
+We log two kinds of messages (this is the extent of the [logr][logr] API):
```

Comment:
I think it's a bit confusing that debug isn't included in this list but is called out in the table above. Can you make it clear that debug is a log level but _not_ a kind?

Reply:
LGTM!